src/bridge/cockpitdbusjson.c

```diff
@@ -691,7 +709,63 @@ build_json_dictionary (const GVariantType *entry_type,
 }
 
 static JsonNode *
-build_json (GVariant *value)
+build_json_fd_channel (GVariant *value,
+                       VariantContext *context)
+{
+  JsonObject *object;
+  GError *error = NULL;
+  JsonNode *node;
+  gint fd = -1;
+  const gchar *old;
+  const gchar *id;
+
+  if (context->fdlist)
+    {
+      fd = g_unix_fd_list_get (context->fdlist, g_variant_get_handle (value), &error);
+      if (fd == -1)
+        {
+          g_warning ("couldn't dup file descriptor from DBus message: %s", error->message);
+          g_clear_error (&error);
+        }
+    }
+
+  if (fd < 0)
+    {
+      node = json_node_new (JSON_NODE_NULL);
+    }
+  else
+    {
+      g_assert (context->fdlist != NULL);
+      g_assert (context->fdids != NULL);
+
+      node = json_node_new (JSON_NODE_OBJECT);
+
+      /* Add a new internal channel name for this file descriptor */
+      id = cockpit_pipe_channel_add_internal_fd (fd);
+      g_queue_push_tail (context->fdids, (gpointer)id);
+
+      /* And only keep the last N ready for opening channels */
+      while (g_queue_get_length (context->fdids) > MAX_RECEIVED_DBUS_FDS)
+        {
+          old = (const gchar *)g_queue_pop_head (context->fdids);
```

Comment:
Client code can decide to ignore a passed fd. So we have to keep the fd open for  a while, so that the cockpit channel can be created if a client does not ignore it. The problem is when to close fds that were ignored. We decided to just keep the last N fds open.

This is why we can't warn about it, because many of the instances are just normal operation. If the limit turns out too low, the client code will get an error when it establishes the channel.